### PR TITLE
[BUG][GUI] Non-static coinControl object in CoinControlDialogs

### PR DESCRIPTION
**Pull request by random-zebra**
This is an alternative to #1691 .

Instead of fixing the the issue by checking for edits to `coinControl` in the show event of the dialogs, make `coinControl` a class variable of `CoinControlDialog`, so each dialog instance has its own independent object, and there are no inconsistencies.

### CODE REVIEW

File: src/qt/pivx/send.cpp [hunk 1]
```diff
@@ -439,27 +439,32 @@ void SendWidget::updateEntryLabels(QList<SendCoinsRecipient> recipients)
...
-    if (openDialogWithOpaqueBackgroundY(dialog, window, 3, 5)) {
-        CTxDestination dest = DecodeDestination(dialog->getAddress().toStdString());
-        // Ask if it's what the user really wants
-        if (!walletModel->isMine(dest) &&
-            !ask(tr("Warning!"), tr("The change address doesn't belong to this wallet.\n\nDo you want to continue?"))) {
+
+    CTxDestination destChange = (openDialogWithOpaqueBackgroundY(dialog, window, 3, 5) ?
+                                 dialog->getDestination() : CNoDestination());
+
+    if (boost::get<CNoDestination>(&destChange)) {
```
Comment (furszy) [inline src/qt/pivx/send.cpp:456]: better to use `IsValidDestination` instead.

**Comment by random-zebra** [comment]
The first commit is needed in order to save/return directly the destination, without having to do the decoding outside the dialog, which, in my opinion, is much more clean than returning a string.

No dependency is added. `standard.h` was already included via `coincontrol.h` in order to use `CTxDestination`, which is just a primitive typedef, does not need a special wrapper in the GUI model.

File: src/qt/pivx/send.cpp (hunk 1 above)
Comment (random-zebra) [reply src/qt/pivx/send.cpp]: :+1: Refactored in a separate commit, along with the other few instances found in the code.

**Comment by furszy** [comment]
`CTxDestination` is a backend primitive. The GUI doesn't need to know about its existence.

We shouldn't be using any backend primitive in the GUI and use only walletModel calls. Be as independent as possible (same position for the `coincontrol` object too. It shouldn't be there neither).

In some future, we should have no `standard.h` dependency anywhere in the GUI. The file was made mostly for scripts and the GUI doesn't not handle scripts, not its responsibility.

**Comment by random-zebra** [comment]
This is not a "backend" primitive, nor a call.
`CTxDestination` is just a type (just a boost::variant between CKeyID and CScriptID).
It is already included and used in many places and (of course) the GUI is aware of it:
- settingsbittoolwidget
- addresseswidget
- receive and send widget
- coldstaking widget

... and the list goes on.
The only reason why you don't see `standard.h` included is because it gets included indirectly via other headers.

I don't see any valid reason to create a wrapper for `CTxDestination` in the model, same as you wouldn't wrap a `CBlock`, a `CTransaction` or a `CScript` object. I agree on the separation between GUI and backend, but this does not mean to create a wrapper for every possible class defined in the software.

**Comment by random-zebra** [comment]
Anyway, even if find a valid reason to remove completely `CTxDestination` (or create a wrapper for it in the model), it would be a bigger refactoring needing its own PR (since it's used everywhere).
The scope of this PR is focused on removing the inconsistencies of a __static__ coin control, accessed by separate widgets.

**Comment by furszy** [comment]
We are actually having a wrapper for `CTransaction`, it's the `TransactionRecord` (which is more an output wrapper with the transaction data). And even with that, we are hiding it from the GUI objects behind the transactionModel. Almost every call looking for any data in the tx is done via the `QModelIndex.data(column_xxx)` to return QT objects and not include any dependency to a backend file anywhere in the GUI objects (this "anywhere" is more a "most of the time"), only include it in the model side. 
There are few data calls that are not done in this way that we should migrate when we can.

Having CTxDestination everywhere in the GUI is actually something that we should improve. As you just said, it's everywhere currently (badly). Not a reason to treat it as a standard and continue adding it in more places.

I'm not talking about a wrapper for this for now. I'm going even simpler, talking about the dialog returning/storing a QString, calling to `walletModel->validateDestination(qstring)` (only place calling to the global `IsDestinationValid()` function), then the caller setting the string into whatever place it's set and done. No need for `CTxDestination`.

In this way we centralize the backend dependencies into one single place and purely use "QT friendly" objects in the GUI. And in some future we will be able to only change an specific set of files (only the four models: wallet, tx, addresses and MN) to decouple GUI from backend.

**Comment by random-zebra** [comment]
> Having CTxDestination everywhere in the GUI is actually something that we should improve.

In that case, as I said, it should be done comprehensively in the whole GUI, and not just in this occurrence Here we just moved the use of `CTxDestination` from the send widgegt to the changeaddress widget. It was already used in the GUI.
We are not "adding it in more places". 
Returning a string from the changeaddress widget to the send widget, and then doing the decoding in the latter does not change anything in that sense (besides making it a bit more ugly in my opinion).

In fact, there are a lot of functions in the wallet model that take a destination as argument (`getKeyCreationTime`, `getLabelForAddress`, `getKeyId`, `isMine`, etc...). Those are meant to be called from the other places in the GUI, passing a CTxDestination to them.

We can rethink this separation in the future, but I wouldn't hold up a **bugfix** PR for this reason.

**Comment by random-zebra** [comment]
Merging...

Sidenote about the use of `CTxDestination`: 
accessing it from the GUI also gives performance improvements. Relying always on a qstring means requiring multiple decoding operations (at least one during `IsValidDestinationString`, when we want to check the validity of a PIVX address, and another one in the model, when the address string is actually decoded into CTxDestination, to be used from the actual backend primitive).

**Pull request merged**